### PR TITLE
Add network entitlements tests

### PR DESCRIPTION
**Pull request by n4ss**
Changes done in this PR:
- `test` rule in the Makefile fixed to include package tests in subdirectories (new tests too)
- add a seccomp default profile for test purposes in **seccomp_default.json** and a profile loader in **defaults/helpers-seccomp.go**
- add helper functions to check enabled/disabled security options in OCI spec in **defaults/helpers.go**
- add `network.*` entitlements tests in **defaults/network-ents_test.go**

Follow-up PR will contain:
- `security.*` & `host.*` entitlements tests

### CODE REVIEW

**Comment by n4ss** [comment]
I'll have to rewrite `seccompSyscallWithArgsBlocked` as it doesn't handle *complex* cases such as several seccomp rule whitelisting one argument at the time.

I propose for this to:
- once:
  - go through the seccomp profile
  - build a map of [syscalls<->whitelisted/blacklisted args]
- lookup in this new map for blocked/allowed syscalls (with args)

File: defaults/helpers.go [hunk 1]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
 
 	return ociProfile, nil
 }
+
+// FIXME factor for all defaults package and even project-wide
+func testSpec() *specs.Spec {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file - or is this intended to be a generic spec to be reused for defaults?

File: defaults/helpers.go [hunk 2]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+		s.Linux.Seccomp = &specs.LinuxSeccomp{}
+	}
+
+	s.Linux.Seccomp = seccomp
+
+	return s
+}
+
+// We can't create a map to check set inclusion with specs.LinuxSeccompArg as key type in Golang so no O(n)
+func seccompArgsContains(syscallArgsSet, syscallArgsSubset []specs.LinuxSeccompArg) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should consider moving it to the test file

File: defaults/helpers.go [hunk 3]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+
+		if !found {
+			return false
+		}
+	}
+
+	return true
+}
+
+func matchSeccompRule(seccompRule specs.LinuxSyscall, syscallName string, syscallArgs []specs.LinuxSeccompArg) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 4]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+			if seccompArgsContains(seccompRule.Args, syscallArgs) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func seccompSyscallWithArgsBlocked(seccompProfile specs.LinuxSeccomp, syscallName types.Syscall, syscallArgs []specs.LinuxSeccompArg) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 5]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+					return true
+				}
+			}
+		}
+	}
+
+	return blocked
+}
+
+func seccompSyscallsWithArgsBlocked(seccompProfile specs.LinuxSeccomp, syscallsWithArgs map[types.Syscall][]specs.LinuxSeccompArg) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 6]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+	for syscallName, syscallArgs := range syscallsWithArgs {
+		if !seccompSyscallWithArgsBlocked(seccompProfile, syscallName, syscallArgs) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func seccompSyscallsBlocked(seccompProfile specs.LinuxSeccomp, syscallNames []types.Syscall) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 7]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+	for _, syscallName := range syscallNames {
+		if !seccompSyscallWithArgsBlocked(seccompProfile, syscallName, []specs.LinuxSeccompArg{}) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func capBlocked(linuxCaps specs.LinuxCapabilities, capability types.Capability) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 8]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+	return !(capListContains(linuxCaps.Bounding, capability) || capListContains(linuxCaps.Permitted, capability) ||
+		capListContains(linuxCaps.Inheritable, capability) || capListContains(linuxCaps.Effective, capability))
+}
+
+func capsBlocked(linuxCaps specs.LinuxCapabilities, capabilities []types.Capability) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 9]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+	for _, capability := range capabilities {
+		if !capBlocked(linuxCaps, capability) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func capAllowed(linuxCaps specs.LinuxCapabilities, capability types.Capability) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 10]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+	for _, capability := range capabilities {
+		if !capAllowed(linuxCaps, capability) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func namespaceActivated(nsList []specs.LinuxNamespace, namespace specs.LinuxNamespaceType) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 11]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+	for _, ns := range nsList {
+		if ns.Type == namespace {
+			return true
+		}
+	}
+
+	return false
+}
+
+func namespacesActivated(nsList []specs.LinuxNamespace, namespaces []specs.LinuxNamespaceType) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: defaults/helpers.go [hunk 12]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+	for _, namespace := range namespaces {
+		if !namespaceActivated(nsList, namespace) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func namespacesDeactivated(nsList []specs.LinuxNamespace, namespaces []specs.LinuxNamespaceType) bool {
```
Comment (riyazdf) [inline defaults/helpers.go]: if this is only used for testing, we should move it to the test file

File: secprofile/oci-profile.go [hunk 13]
```diff
@@ -111,8 +111,6 @@ func allowSyscallWithArgs(seccompProfile *specs.LinuxSeccomp, syscallName types.
 
 	syscallNameToAllowStr := string(syscallName)
 
-	logrus.Errorf("Allowing syscall: %s", syscallNameToAllowStr)
```
Comment (riyazdf) [inline secprofile/oci-profile.go]: why remove this?  It may be helpful as a debug level

File: secprofile/oci-profile.go [hunk 14]
```diff
@@ -134,8 +132,6 @@ func allowSyscallWithArgs(seccompProfile *specs.LinuxSeccomp, syscallName types.
 			Args:   syscallArgs,
 		}
 		seccompProfile.Syscalls = append(seccompProfile.Syscalls, newRule)
-
-		logrus.Errorf("Adding allow rule for syscall: %s", syscallNameToAllowStr)
```
Comment (riyazdf) [inline secprofile/oci-profile.go]: same comment about keeping as debug

File: defaults/helpers.go [hunk 15]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+
+	s.Linux.Seccomp = seccomp
+
+	return s
+}
+
+// We can't create a map to check set inclusion with specs.LinuxSeccompArg as key type in Golang so no O(n)
+func seccompArgsContains(syscallArgsSet, syscallArgsSubset []specs.LinuxSeccompArg) bool {
+	if len(syscallArgsSet) < len(syscallArgsSubset) ||
+		(len(syscallArgsSubset) == 0 && len(syscallArgsSet) != 0) {
```
Comment (riyazdf) [inline defaults/helpers.go]: for `len(syscallArgsSubset) == 0 && len(syscallArgsSet) !=0` - isn't that a true case?  Aren't we trying to check if the `syscallArgsSet` contains all of the `syscallArgsSubset`?

Side note: might help to rename vars to better differentiate the master set vs. the subset to check for

File: defaults/helpers.go (hunk 7 above)
Comment (riyazdf) [inline defaults/helpers.go]: I'm wondering if this should be configurable to also account for ambient caps?  Docker doesn't use them as of current, but it would be useful for the future / other workloads.

File: defaults/helpers.go (hunk 9 above)
Comment (riyazdf) [inline defaults/helpers.go]: Same as above re: ambient caps

File: helpers_test.go [hunk 16]
```diff
@@ -0,0 +1,34 @@
+package libentitlement
+
+import (
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func testSpec() *specs.Spec {
```
Comment (riyazdf) [inline helpers_test.go]: is this necessary given the same function exists in helpers.go?  We should remove one.

File: defaults/network-ents_test.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+	"testing"
+
+	"github.com/docker/libentitlement/secprofile"
+	"github.com/docker/libentitlement/types"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNetworkNoneEntitlementEnforce(t *testing.T) {
+	entitlementID := "network.none"
```
Comment (riyazdf) [inline defaults/network-ents_test.go]: we should use the const we have for this: https://github.com/docker/libentitlement/blob/master/defaults/network-ents.go#L18

File: defaults/network-ents_test.go [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+				Op:    specs.OpEqualTo,
+				Value: syscall.AF_LOCAL,
+			},
+		},
+	}
+	require.False(t, seccompSyscallsWithArgsBlocked(*newOCIProfile.OCI.Linux.Seccomp, syscallsWithArgsToAllow))
+}
+
+func TestNetworkUserEntitlementEnforce(t *testing.T) {
+	entitlementID := "network.user"
```
Comment (riyazdf) [inline defaults/network-ents_test.go]: same comment re: const for ID

File: defaults/helpers-seccomp.go [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+	"encoding/json"
+	"fmt"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"io/ioutil"
+)
+
+func setDefaultSeccompProfile() (*specs.LinuxSeccomp, error) {
+	var profile specs.LinuxSeccomp
+
+	jsonProfile, err := ioutil.ReadFile("seccomp_default.json")
```
[the remark this answers is not in the document]
Comment (n4ss) [reply defaults/helpers-seccomp.go]: Will fix, good catch.

File: defaults/network-ents_test.go [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+				Value: syscall.AF_UNIX,
+			},
+			{
+				Index: 0,
+				Op:    specs.OpEqualTo,
+				Value: syscall.AF_LOCAL,
+			},
+		},
+	}
+	require.False(t, seccompSyscallsWithArgsBlocked(*newOCIProfile.OCI.Linux.Seccomp, syscallsWithArgsToAllow))
```
Comment (riyazdf) [inline defaults/network-ents_test.go]: in a follow-up PR: it would be awesome to have an integration test that runs a container similar to how you've demoed this functionality.

File: defaults/helpers.go [hunk 21]
```diff
@@ -4,8 +4,23 @@ import (
 	"fmt"
 
 	"github.com/docker/libentitlement/secprofile"
+	"github.com/docker/libentitlement/types"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"reflect"
 )
 
+func capListContains(capList []string, capability types.Capability) bool {
```
[the remark this answers is not in the document]
Comment (n4ss) [reply defaults/helpers.go]: Hmm capList is actually an OCI capability list.. it should already be validated I guess if filed by libentitlement.. wdyt?

File: defaults/network-ents_test.go [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+
+	newOCIProfile, err := ociProfileConversionCheck(newProfile, NetworkUserEntFullID)
+	require.NoError(t, err, "Failed converting to OCI profile while testing entitlement %s", entitlementID)
+
+	require.NotNil(t, newOCIProfile.OCI)
+	require.NotNil(t, newOCIProfile.OCI.Linux)
+
+	require.NotNil(t, newOCIProfile.OCI.Process.Capabilities)
+
+	capsToRemove := []types.Capability{CapNetAdmin, CapNetBindService, CapNetRaw}
```
Comment (riyazdf) [inline defaults/network-ents_test.go]: can we also test that the correct network caps are included?

File: defaults/helpers.go (hunk 1 above)
Comment (n4ss) [reply defaults/helpers.go]: this is for test purposes only, I'll move this in a test file

File: defaults/network-ents_test.go [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+	require.True(t, capsBlocked(*newOCIProfile.OCI.Process.Capabilities, capsToRemove))
+
+	nsToAdd := []specs.LinuxNamespaceType{specs.NetworkNamespace}
+	require.True(t, namespacesActivated(newOCIProfile.OCI.Linux.Namespaces, nsToAdd))
+}
+
+func TestNetworkProxyEntitlementEnforce(t *testing.T) {
+	entitlementID := "network.proxy"
```
Comment (riyazdf) [inline defaults/network-ents_test.go]: same comment re: const for ID

File: defaults/helpers.go [hunk 24]
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
...
+				CPU:     &specs.WindowsCPUResources{},
+				Storage: &specs.WindowsStorageResources{},
+				Network: &specs.WindowsNetworkResources{},
+			},
+		},
+	}
+
+	seccomp, err := setDefaultSeccompProfile()
+	if err != nil {
+		// In case we get an error before seccomp struct is fully updated from decoded json, we empty it manually.
```
[the remark this answers is not in the document]
Comment (n4ss) [reply defaults/helpers.go]: Right, will fix.

File: defaults/helpers.go (hunk 2 above)
Comment (n4ss) [reply defaults/helpers.go]: totally, actually wanted to do that in the first place 😖 will fix

File: defaults/helpers.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (riyazdf) [reply defaults/helpers.go]: ok that's fine with me.  This is an internal helper for testing in any case.

File: defaults/helpers.go (hunk 15 above)
Comment (n4ss) [reply defaults/helpers.go]: I need to rename this function or make a wrapper around it because it's currently doing more than just checking if `syscallArgsSubset` is contained in `syscallArgsSet`.

Answer to the 1st question:
Yes in theory.. but..
No because `syscallArgSubset` contains the syscall args of the to-be-checked syscall. If there are no arguments, that means that we want to completely whitelist the syscall completely so the whitelisting rule should be absolute (hence no args either).

Not sure I'm clear there.. 
function should actually be called `seccompArgsMatched(syscallRuleArgs, syscallArgsToCheck..`

File: defaults/helpers.go (hunk 7 above)
Comment (n4ss) [reply defaults/helpers.go]: [As discussed] that makes sense but for now I don't want false positive on Moby. I have to talk with the runc people to understand the tradeoffs, who might use ambient caps and why.

File: secprofile/oci-profile.go (hunk 13 above)
Comment (n4ss) [reply secprofile/oci-profile.go]: I felt like this info wasn't very useful as its pretty much repeats what the function is doing. Will undo that and switch it to debug if you think we should keep it.

File: helpers_test.go (hunk 16 above)
Comment (n4ss) [reply helpers_test.go]: [As discussed] I'll move this as well as *shareable* helper code for test purposes to a `testutil` package.

File: secprofile/oci-profile.go (hunk 13 above)
Comment (riyazdf) [reply secprofile/oci-profile.go]: ok.  Let's leave this out and create a tracking issue for logging in libentitlement.  If there is enough demand, we can add it over the core codebase all at once

File: defaults/network-ents_test.go (hunk 17 above)
Comment (n4ss) [reply defaults/network-ents_test.go]: indeed, will fix for all.

File: defaults/network-ents_test.go (hunk 20 above)
Comment (n4ss) [reply defaults/network-ents_test.go]: This will very probably be done on the Moby side but this is the next thing on my todo-list.

File: secprofile/oci-profile.go (hunk 13 above)
Comment (n4ss) [reply secprofile/oci-profile.go]: Sounds good! Issue here: https://github.com/docker/libentitlement/issues/46

File: testutils/ocispec.go [hunk 25]
```diff
@@ -0,0 +1,219 @@
+package testutils
+
+import (
+	"github.com/docker/libentitlement/types"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"reflect"
+)
+
+var (
+	testCapslist = mobyDefaultCaps
```
Comment (riyazdf) [inline testutils/ocispec.go]: I don't think this is used, should it be removed?  As discussed, it would be good to use `mobyDefaultCaps` in tests such that we check an entitlement's cap list is:

`mobyDefaultCaps + entitlement.capsToAdd - entitlement.capsToRemove`

**Comment by n4ss** [comment]
New changes:
- moved test helpers, test profiles and utils structs to `testutils` package
- moved typed capabilities and syscalls to `osdefs` package, this will be shareable by OCI test helpers in `testutils` and actual entitlements (and tests) in `defaults` package
- check perfect matching between `test struct's capabilities + caps to add for entitlement - caps to remove for entitlement` and what we obtain when enforcing the entitlement separately with `OCICapsMatchRefWithConstraints`
- fix bugs at https://github.com/docker/libentitlement/pull/45/commits/add301075b96db1a7dd5fb0bc40c915cf6ec7e03#diff-d79a0466acbe1316a0ea9bb98db1607bR56 and https://github.com/docker/libentitlement/pull/45/commits/add301075b96db1a7dd5fb0bc40c915cf6ec7e03#diff-d79a0466acbe1316a0ea9bb98db1607bR53

File: testutils/ocispec.go [hunk 26]
```diff
@@ -0,0 +1,279 @@
...
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// SeccompSyscallWithArgsBlocked checks that the provided syscall and args are blocked by the seccomp profile
+func SeccompSyscallWithArgsBlocked(seccompProfile specs.LinuxSeccomp, syscallName types.Syscall, syscallArgs []specs.LinuxSeccompArg) bool {
```
Comment (ashfall) [inline testutils/ocispec.go]: naming nit: Can we rename this to indicate that this is checking something? `SeccompSyscallWithArgsBlocked` doesn't tell us if it's a checker. For example, `IsSyscallBlockedBySeccomp` or `IsSyscallWithArgsBlockedBySeccomp` is a more descriptive name.

File: testutils/ocispec.go [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+				}
+			}
+		}
+	}
+
+	return blocked
+}
+
+// SeccompSyscallsWithArgsBlocked checks that the provided list of syscalls and args are blocked by the seccomp profile
+func SeccompSyscallsWithArgsBlocked(seccompProfile specs.LinuxSeccomp, syscallsWithArgs map[types.Syscall][]specs.LinuxSeccompArg) bool {
```
Comment (ashfall) [inline testutils/ocispec.go]: See above re:naming

File: testutils/ocispec.go [hunk 28]
```diff
@@ -0,0 +1,279 @@
...
+		if !SeccompSyscallWithArgsBlocked(seccompProfile, syscallName, syscallArgs) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// SeccompSyscallsBlocked checks that the provided syscalls are blocked by the seccomp profile
+func SeccompSyscallsBlocked(seccompProfile specs.LinuxSeccomp, syscallNames []types.Syscall) bool {
```
Comment (ashfall) [inline testutils/ocispec.go]: See above re:naming

File: testutils/ocispec.go [hunk 29]
```diff
@@ -0,0 +1,279 @@
...
+	s.Process.Capabilities.Inheritable = getDefaultCapList()
+	s.Process.Capabilities.Permitted = getDefaultCapList()
+
+	return s
+}
+
+// SyscallArgsMatchSeccompRuleArgs checks that the seccomp rule's args match the provided syscall args
+func SyscallArgsMatchSeccompRuleArgs(syscallArgsRule, syscallArgsSyscall []specs.LinuxSeccompArg) bool {
+	if len(syscallArgsRule) < len(syscallArgsSyscall) ||
+		(len(syscallArgsSyscall) == 0 && len(syscallArgsRule) != 0) {
```
Comment (ashfall) [inline testutils/ocispec.go]: The `syscallArgs` in `syscallArgsSyscall` and `syscallArgsRule` seems redundant, perhaps we can rename to `argsFromSyscall` and `argsFromRule`?

File: testutils/ocispec.go [hunk 30]
```diff
@@ -0,0 +1,279 @@
...
+		return false
+	}
+
+	for _, syscallArgFromSyscall := range syscallArgsSyscall {
+		found := false
+
+		for _, syscallArgsFromRule := range syscallArgsRule {
```
Comment (ashfall) [inline testutils/ocispec.go]: `syscallArgsFromRule` sounds like a list -is this intentional or should this be `syscallArgFromRule`?

File: testutils/ocispec.go [hunk 31]
```diff
@@ -0,0 +1,279 @@
...
+		if ns.Type == namespace {
+			return true
+		}
+	}
+
+	return false
+}
+
+// NamespacesActivated checks that the namespaces in the provided ns list are enabled
+func NamespacesActivated(nsList []specs.LinuxNamespace, namespaces []specs.LinuxNamespaceType) bool {
```
Comment (ashfall) [inline testutils/ocispec.go]: naming nit: Probably `areNamespacesActivated` or `checkNamespacesDeactivation`?

File: testutils/ocispec.go [hunk 32]
```diff
@@ -0,0 +1,279 @@
...
+
+	for _, capStrList := range capStrLists {
+		match = match && capsListMatchRefWithConstraints(capStrList, capsToAdd, capsToRemove)
+	}
+
+	return match
+}
+
+// NamespaceActivated checks that the provided namespace is enabled
+func NamespaceActivated(nsList []specs.LinuxNamespace, namespace specs.LinuxNamespaceType) bool {
```
[the remark this answers is not in the document]
Comment (ashfall) [reply testutils/ocispec.go]: Also, naming nit: `isNamespaceActivated`

File: testutils/ocispec.go [hunk 33]
```diff
@@ -0,0 +1,279 @@
...
+		if !NamespaceActivated(nsList, namespace) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// NamespacesDeactivated checks that the namespaces in the provided ns list are disabled
+func NamespacesDeactivated(nsList []specs.LinuxNamespace, namespaces []specs.LinuxNamespaceType) bool {
```
Comment (ashfall) [inline testutils/ocispec.go]: nit: `areNamespacesDeactivated` or `checkNamespacesDeactivation`

**Comment by n4ss** [comment]
> Looks like most of the test helpers are not being used outside the package, is there a reason we are defining them as exported functions? If not, let's default to unexported test helpers, and we can always export/move them if we need them elsewhere. Thoughts?

Yep, sounds good. Will fix.

> Can we decide on a convention for helpers that return a bool in that their name should indicate so? fooStatusChecker or isFooActive and so on instead of fooActivated or fooDeactivated

Good catch, `isFooActive` works for me. Will fix.

File: testutils/ocispec.go (hunk 29 above)
Comment (n4ss) [reply testutils/ocispec.go]: Yep!

File: testutils/ocispec.go (hunk 30 above)
Comment (n4ss) [reply testutils/ocispec.go]: Will fix!

File: testutils/ocispec.go (hunk 26 above)
Comment (n4ss) [reply testutils/ocispec.go]: Yeah I was wondering if the "is" was necessary and indeed it seems weird without it. Will fix all the names.

**Comment by riyazdf** [comment]
following up on naming: @ashfall and I were discussing possible different naming conventions for the matching functions like `capsListMatchRefSet` and `OCICapsMatchRefWithConstraints`.

These functions are really checking for set equivalence, perhaps `hasEquivalentCaps` and `HasEquvalentOCICaps`? I'd also be fine with creating a tracking issue for this and punting until we have more consumers of this library give feedback.

File: Makefile [hunk 34]
```diff
@@ -18,4 +18,4 @@ lint:
 
 
 test: lint
-	go test .
```
Comment (riyazdf) [inline Makefile]: we should add codecov to this.  Can be a separate PR / tracking issue.

File: testutils/moby-defaults.go [hunk 35]
```diff
@@ -0,0 +1,77 @@
+package testutils
+
+import (
+	"encoding/json"
```
Comment (riyazdf) [inline testutils/moby-defaults.go]: nit: can we put the system imports separately at the top?

Comment (n4ss) [reply testutils/moby-defaults.go]: will do, gofmt actually shuffled everything..

File: Makefile (hunk 34 above)
Comment (n4ss) [reply Makefile]: will do

**Comment by n4ss** [comment]
@riyazdf I'd prefer to do that in a follow-up PR, I'm gonna submit my second PR for the rest of the entitlement testing and then we can discuss more naming conventions. We can also discuss naming conventions non-test methods there. Sounds good?
